sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/decommission/DecommissionPlanFactory.java

```diff
@@ -0,0 +1,289 @@
+                    stateStore.storeGoalOverrideStatus(task.getName(), GoalStateOverride.Status.INACTIVE);
+                }
+            }
+        }
+
+        if (podsToDecommission.isEmpty()) {
+            return new PlanInfo(Optional.empty(), Collections.emptyList());
+        }
+
+        Collection<Step> resourceSteps = new ArrayList<>();
+        List<Phase> phases = new ArrayList<>();
+        // Each pod to be decommissioned gets its own phase in the decommission plan:
+        for (Map.Entry<PodKey, Collection<Protos.TaskInfo>> entry : podsToDecommission.entrySet()) {
+            List<Step> steps = new ArrayList<>();
+
+            // 1. Kill pod's tasks
+            steps.addAll(entry.getValue().stream()
+                    .map(task -> new TriggerDecommissionStep(stateStore, taskKiller, task))
+                    .collect(Collectors.toList()));
+
+            // 2. Unreserve pod's resources
+            // Note: Even though this step is in a serial phase, in practice resource cleanup should be done in
+            // parallel, as all the tasks had been flagged for decommissioning via TriggerDecommissionStep.
+            Collection<ResourceCleanupStep> resourceStepsForPod =
+                    ResourceUtils.getResourceIds(ResourceUtils.getAllResources(entry.getValue())).stream()
+                            .map(resourceId -> new ResourceCleanupStep(resourceId, Status.PENDING))
+                            .collect(Collectors.toList());
+            resourceSteps.addAll(resourceStepsForPod);
+            steps.addAll(resourceStepsForPod);
+
+            // 3. Delete pod's tasks from ZK
+            // Note: As a side effect, this will also clear the override status.
+            steps.addAll(entry.getValue().stream()
+                    .map(task -> new EraseTaskStateStep(stateStore, task.getName()))
+                    .collect(Collectors.toList()));
+
+            phases.add(new DefaultPhase(
+                    entry.getKey().getPodName(), steps, new SerialStrategy<>(), Collections.emptyList()));
+        }
+
+        return new PlanInfo(Optional.of(new DefaultPlan(Constants.DECOMMISSION_PLAN_NAME, phases)), resourceSteps);
+    }
+
+    /**
+     * A utility class for ordering a {@link SortedMap} of pods to be decommissioned.
+     */
+    @VisibleForTesting
+    static class PodKey implements Comparable<PodKey> {
+        private final int podTypeIndex;
+        private final String podType;
+        private final int podIndex;
+
+        /**
+         * @param taskInfo a taskinfo whose pod is to be sorted
+         * @param orderedPodTypes the list of known pod types in order of priority (reverse of ServiceSpec ordering)
+         */
+        PodKey(String podType, int podIndex, List<String> orderedPodTypes) throws TaskException {
+            this.podType = podType;
+            // If the pod is unknown, it gets a -1, placing it at higher priority than listed pods:
```

Comment:
When would the podtype be unknown? When the pod is dropped from the ServiceSpec during an upgrade?

Reply:
Supporting dropping pod types is very desirable.